### PR TITLE
Plane:add swarm failsafe

### PR DESCRIPTION
**Pull request by XuZeng931201**
add awarm failsafe.

### CODE REVIEW

**Comment by XuZeng931201** [comment]
ask to merge.I test in mavproxy with sitl

**Comment by OXINARF** [comment]
Hi @XuZeng931201, thank you for your first contribution!

Please read our contribution rules (http://ardupilot.org/dev/docs/submitting-patches-back-to-master.html), specially the style guide - you are using tabs instead of spaces for example.

Also, it's likely Plane maintainers are going to ask this parameter to be on G2 (https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/Parameters.cpp#L1104) instead of using a reserved one.

File: ArduPlane/system.cpp [hunk 1]
```diff
@@ -526,6 +526,60 @@ void Plane::exit_mode(enum FlightMode mode)
...
+void Plane::check_swarm_failsafe()
+{
+	uint32_t tnow = millis();
+	int16_t failsafe_counts = 0;
+	// exit if not in guided mode or rtl mode
+	// TODO:We can consider the sitution in auto mode or other position control mode
+	if (!g.swarm_failsafe_enable || (control_mode != GUIDED && control_mode != RTL)) {
+		return;
+	}
+	if (g.throttle_fs_enabled) {
```
Comment (magicrub) [inline ArduPlane/system.cpp]: lots of redundant checking here just to make it ++ vs +=2. How about 
```
if (g.throttle_fs_enabled) {
    failsafe_count_inc = 1;
else {
    failsafe_count_inc = 2;
}

```
and then have the other code say
`failsafe_counts += failsafe_counts_inc
`

File: ArduPlane/Parameters.h [hunk 2]
```diff
@@ -344,7 +344,8 @@ class Parameters {
 
         k_param_DataFlash = 253, // Logging Group
 
-        // 254,255: reserved
+		k_param_swarm_failsafe_enable,
+		// 255: reserved
```
Comment (magicrub) [inline ArduPlane/Parameters.h]: as @OXINARF said, please use g2 and keep these reserved.

Comment (XuZeng931201) [reply ArduPlane/Parameters.h]: when I used astyle --options=Tools/CodeStyle/astylerc <filename>.It shows many formatting changes.I don't know whether is ok for this commit.If something wrong with this.I really feel apologetic about that.

Comment (XuZeng931201) [reply ArduPlane/Parameters.h]: when I used astyle --options=Tools/CodeStyle/astylerc <filename>.It shows many formatting changes.I don't know whether is ok for this commit.If something wrong with this.I really feel apologetic about that.

**Comment by magicrub** [comment]
Well, the code change looks better. Thank you. Please please please make whitespace changes on a separate commit. Would you like me to fix that for you?

**Comment by WickedShell** [comment]
Why call this a swam failsafe? Seems to be just a failsafe for guided only and a failsafe of if you have both a RC and GCS failsafe? Surely that could be done more generically then calling it swam failsafe? And the specificness to only RTL seems a bit out of place, surely it should have a selectable actions list?

File: ArduPlane/Plane.h [hunk 3]
```diff
@@ -1079,12 +1083,12 @@ class Plane : public AP_HAL::HAL::Callbacks {
 
     // list of priorities, highest priority first
     static constexpr int8_t _failsafe_priorities[] = {
-                                                      Failsafe_Action_Terminate,
-                                                      Failsafe_Action_Land,
-                                                      Failsafe_Action_RTL,
-                                                      Failsafe_Action_None,
-                                                      -1 // the priority list must end with a sentinel of -1
-                                                     };
+        Failsafe_Action_Terminate,
```
Comment (WickedShell) [inline ArduPlane/Plane.h]: Bad whitespace change

**Comment by WickedShell** [comment]
There are a number of bad changes in the whitespace changes....

**Comment by magicrub** [comment]
> Why call this a swam failsafe?
Agreed. The name needs to be more generic. It is not specific to swarming at all

**Pull request closed without merging**